packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts

```diff
@@ -156,6 +156,13 @@ export interface CommonAutoScalingGroupProps {
    * @default none
    */
   readonly spotPrice?: string;
+
+  /**
+   * Configuration for health checks
+   *
+   * @default - HealthCheckConfiguration with defaults.
+   */
+  readonly healthCheck?: HealthCheckConfiguration;
```

Comment:
This will lends itself nicely to a "union-like" API:

```ts
export interface Ec2HealthCheckOptions {
  readonly grace?: Duration;
}

export interface ElbHealthCheckOptions {
  readonly grace: Duration; // <-- required here
}

export class HealthCheck {
  public static ec2(options: Ec2HealthCheckOptions = {}): HealthCheck {
    return new HealthCheck('EC2', options.grace && options.grace.toSeconds());
  }

  public static elb(options: ElbHealthCheckOptions): HealthCheck {
    return new HealthCheck('ELB', options.grace.toSeconds());
  }

  private constructor(public readonly type: string, public readonly gradePeriod?: number) { }
}
```

And then the usage is:

```ts
healthCheck: HealthCheck.ec2();
healthCheck: HealthCheck.ec2({ grace: Duration.minutes(5) });
healthCheck: HealthCheck.elb({ grace: Duration.minutes(5) }); // <-- required grace enforced by compiler
```

Reply:
That's a much cleaner API, thanks.

How would you feel about using a `Partial<HealthCheckOptions>` instead of repeating the Options Interface? 

```ts
export interface HealthCheckOptions {
  readonly grace: Duration;
}

export class HealthCheck {
  public static ec2(options: Partial<HealthCheckOptions> = {}): HealthCheck {
    return new HealthCheck('EC2', options.grace && options.grace.toSeconds());
  }

  public static elb(options: HealthCheckOptions): HealthCheck {
    return new HealthCheck('ELB', options.grace.toSeconds());
  }

  private constructor(public readonly type: string, public readonly gradePeriod?: number) { }
}
```

Reply:
Never mind, it seems like JSII doesn't support the Partial keyword:
```
09:46:31 - File change detected. Starting incremental compilation...
@aws-cdk/aws-autoscaling: lib/auto-scaling-group.ts:712:21 - error TS9999: JSII: Only string index maps are supported
@aws-cdk/aws-autoscaling: 712   public static ec2(options: Partial<HealthCheckOptions> = {}): HealthCheck {
@aws-cdk/aws-autoscaling:                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Reply:
Done, thanks